merkletree.go

```diff
@@ -0,0 +1,275 @@
+package merkletree
+
+import (
+	"errors"
+	"hash"
+)
+
+var (
+	ErrInvalidTree = errors.New("[merkletree] invalid tree")
+)
+
+const (
+	EmptyBranchIdentifier = 'E'
+	LeafIdentifier        = 'L'
+)
+
+type HashSuite struct {
+	HashSizeByte int
+	Hash         hash.Hash
+}
+
+type MerkleTree struct {
+	hash      HashSuite
+	treeNonce []byte
+	salt      []byte
+	root      *interiorNode
+}
+
+type interiorNode struct {
+	parent     *interiorNode
+	leftChild  interface{}
+	rightChild interface{}
+	leftHash   []byte
+	rightHash  []byte
+	level      int
+}
+
+type userLeafNode struct {
+	interiorNode
+	key   string
+	value []byte
+	index []byte
```

Comment:
The field for the cryptographic commitment ```commit(key || value)``` should still be included since it needs to be returned as part of the auth path.